### ISSUE TITLE
1.0.0 RC2/RC4: Could not determine my public IPv4 address

### ISSUE THREAD

**Issue by sanderjo** [sabnzbd/sabnzbd#498]
With 1.0.0 RC2 and RC4 I get "Could not determine my public IPv4 address"

RC1 is still OK.

I will investigate and report back. Related to the move of the ip address discovery to a different location?

I only discovered this now because @jcfp PPA still seems to be on RC1, so that is still on my Ubuntu/Debian boxes ...

```
2016-03-05 10:50:12,678::INFO::[SABnzbd:1244] SABnzbd.py-1.0.0RC4 (rev=1ba99769794134b18c8416631d51283ac4409b88)
2016-03-05 10:50:12,679::INFO::[SABnzbd:1245] Full executable path = /home/sander/sab-1.0.0-rc4/SABnzbd-1.0.0RC4/SABnzbd.py
2016-03-05 10:50:12,679::INFO::[SABnzbd:1257] Platform = posix
2016-03-05 10:50:12,680::INFO::[SABnzbd:1258] Python-version = 2.7.9 (default, Apr  2 2015, 15:33:21) 
[GCC 4.9.2]
2016-03-05 10:50:12,680::INFO::[SABnzbd:1259] Arguments = ./SABnzbd.py
2016-03-05 10:50:12,680::DEBUG::[SABnzbd:1266] My local IPv4 address = 192.168.1.245
2016-03-05 10:50:12,943::DEBUG::[SABnzbd:1274] Could not determine my public IPv4 address
2016-03-05 10:50:12,944::DEBUG::[SABnzbd:1278] My IPv6 address = 2001:470:blabla
```

**Comment by Safihre** [comment]
Works fine here on Ubuntu 15.10 with RC4:
```
2016-03-05 10:11:44,502::INFO::[SABnzbd:1243] --------------------------------
2016-03-05 10:11:44,502::INFO::[SABnzbd:1244] SABnzbd.py-1.0.0RC4 (rev=1ba99769794134b18c8416631d51283ac4409b88)
2016-03-05 10:11:44,502::INFO::[SABnzbd:1245] Full executable path = /home/osboxes/Documents/SABnzbd-1.0.0RC4/SABnzbd.py
2016-03-05 10:11:44,503::INFO::[SABnzbd:1257] Platform = posix
2016-03-05 10:11:44,503::INFO::[SABnzbd:1258] Python-version = 2.7.10 (default, Oct 14 2015, 16:09:02) 
[GCC 5.2.1 20151010]
2016-03-05 10:11:44,503::INFO::[SABnzbd:1259] Arguments = SABnzbd.py
2016-03-05 10:11:44,503::DEBUG::[SABnzbd:1266] My local IPv4 address = 10.0.bla.bla
2016-03-05 10:11:44,846::DEBUG::[SABnzbd:1272] My public IPv4 address = 46.126.bla.bla
2016-03-05 10:11:44,846::DEBUG::[SABnzbd:1280] Could not determine my IPv6 address
2016-03-05 10:11:44,856::DEBUG::[SABnzbd:1286] CPU Pystone available performance is 162892
```

**Comment by sanderjo** [comment]
Ah, found the problem: When I visit http://self-test.sabnzbd.org/ ... that site tells me my IPv6 address (I'm both IPv4 & IPv6 connected) ... which of course leads to problems with ```socket.inet_aton(public_ipv4)```

Let's see if we can force ```urllib2.Request()``` to use IPv4...

**Comment by Safihre** [comment]
Should have thought of that.. Server can't see/return IPv4 when connected with IPv6 of course.
http://api.ipify.org/ only allowed IPv4 connections, but we want all in 1 of course.

**Comment by Safihre** [comment]
Can't you force IPv4 just like you do here?
https://github.com/sabnzbd/sabnzbd/blob/develop/sabnzbd/getipaddress.py#L29

**Comment by sanderjo** [comment]
I don't think so. I have something else brewing. BRB

**Comment by Safihre** [comment]
I only saw your message now and used this solution from http://stackoverflow.com/a/18007585
```python
def publicipv4():
    try:
        import urllib2
        # Force IPv4 (gethostbyname doesn't do IPv6)
        host = socket.gethostbyname(sabnzbd.cfg.selftest_host())
        req = urllib2.Request('http://{}'.format(host),
                          headers = {'Host': 'self-test.sabnzbd.org', 'User-Agent': 'SABnzbd+/%s' % sabnzbd.version.__version__})
        f = urllib2.urlopen(req, timeout=2)    # timeout 2 seconds, in case website is not accessible
        public_ipv4 = f.read()
        socket.inet_aton(public_ipv4)  # if we got anything else than a plain IPv4 address, this will raise an exception
    except:
        public_ipv4 = None

    return public_ipv4
```

**Comment by sanderjo** [comment]
```
$ host self-test.sabnzbd.org
self-test.sabnzbd.org has address 104.18.41.215
self-test.sabnzbd.org has address 104.18.40.215
self-test.sabnzbd.org has IPv6 address 2400:cb00:2048:1::6812:28d7
self-test.sabnzbd.org has IPv6 address 2400:cb00:2048:1::6812:29d7
```
but
```
$ python -c "import socket; print socket.gethostbyname('self-test.sabnzbd.org')"
104.18.40.215
```
... so only 1 IPv4 address. That is the same ugly part as in my code: what if that specific IPv4 is down? With this code the other IPv4 address(es) are not tried. :-(

**Comment by sanderjo** [comment]
Based on https://stackoverflow.com/questions/2014534/force-python-mechanize-urllib2-to-only-use-a-requests/6319043#6319043 I wrote some testing code:

```
selftest_host = 'self-test.sabnzbd.org'
import socket, urllib2

req = urllib2.Request("http://" + selftest_host, headers={'User-Agent': 'SABnzbd+/1.2.3' , 'Host':'self-test.sabnzbd.org'} )
f = urllib2.urlopen(req, timeout=2)    # timeout 2 seconds, in case website is not accessible
public_ip = f.read()
print "public IP", public_ip


# only resolve IPv4, then connect to the first IPv4 found:
result = socket.getaddrinfo(selftest_host, 80, socket.AF_INET)
selftest_ipv4 = result[0][4][0]		# ... we should also handle the other IPv4 addresses ...
req = urllib2.Request("http://" + selftest_ipv4, headers={'User-Agent': 'SABnzbd+/1.2.3' , 'Host':'self-test.sabnzbd.org'} ) 
f = urllib2.urlopen(req, timeout=2)    # timeout 2 seconds, in case website is not accessible
public_ipv4 = f.read()
print "Public IPv4", public_ipv4


# Now the more elegant stuff: 


# keep it
import socket
origGetAddrInfo = socket.getaddrinfo
# redefine:
def getAddrInfoWrapper(host, port, family=0, socktype=0, proto=0, flags=0):
    return origGetAddrInfo(host, port, socket.AF_INET, socktype, proto, flags)
# replace the original socket.getaddrinfo by our version
socket.getaddrinfo = getAddrInfoWrapper
req = urllib2.Request("http://" + selftest_host, headers={'User-Agent': 'SABnzbd+/1.2.3' , 'Host':'self-test.sabnzbd.org'} ) 
f = urllib2.urlopen(req, timeout=2)    # timeout 2 seconds, in case website is not accessible
public_ipv4 = f.read()
print "Public IPv4", public_ipv4
# restore original resolver:
socket.getaddrinfo = origGetAddrInfo

# Proof it works:
req = urllib2.Request("http://" + selftest_host, headers={'User-Agent': 'SABnzbd+/1.2.3' , 'Host':'self-test.sabnzbd.org'} ) 
f = urllib2.urlopen(req, timeout=2)    # timeout 2 seconds, in case website is not accessible
public_ip = f.read()
print "public IP", public_ip

```
which works on my system:
````
$ python p4.py 
public IP 2001:4c38:blabla
Public IPv4 62.45.bla
Public IPv4 62.45.bla
public IP 2001:4c38:blabla
```
Cool.

So the relevant code is this:
```
# keep it
import socket
origGetAddrInfo = socket.getaddrinfo
# redefine:
def getAddrInfoWrapper(host, port, family=0, socktype=0, proto=0, flags=0):
    return origGetAddrInfo(host, port, socket.AF_INET, socktype, proto, flags)
# replace the original socket.getaddrinfo by our version
socket.getaddrinfo = getAddrInfoWrapper
req = urllib2.Request("http://" + selftest_host, headers={'User-Agent': 'SABnzbd+/1.2.3' , 'Host':'self-test.sabnzbd.org'} ) 
f = urllib2.urlopen(req, timeout=2)    # timeout 2 seconds, in case website is not accessible
public_ipv4 = f.read()
print "Public IPv4", public_ipv4
# restore original resolver:
socket.getaddrinfo = origGetAddrInfo
```
Now the question: do we want all this code in SABnzbd? Or are there easier, more elegant ways?

**Comment by sanderjo** [comment]
BTW ... ;-)

![copying and pasting from stack overflow](https://cloud.githubusercontent.com/assets/1273502/13550371/8c154906-e31c-11e5-9410-071762da2ce8.png)

**Comment by Safihre** [comment]
The url and thus also the IP addresses are backed by Cloudflare (click: http://104.18.41.215), so I don't think availability of the specific ip should be an issue? 
So I would opt for the simpler solution, since it's functioning is not vital to sabnzbd's functioning. 
What do you think?

**Comment by sanderjo** [comment]
http://104.18.41.215/ does not work. Cause: more than one webserver on one IP address. Therefore you need to specify the Host as part of the HTTP request.

Furthermore: just trying one iPv4 will often work, but not always. The goal of having >1 IPv4 address is reliability (and load balancing). One IPv4 address can be unreachable, but you can still reach another IPv4 address.


I wrote this code. IMHo it is easier to read/understand/maintain:
```
selftest_host = 'self-test.sabnzbd.org'
import socket, urllib2

# we only want IPv4 resolving, so socket.AF_INET:
result = socket.getaddrinfo(selftest_host, 80, socket.AF_INET, 0, socket.IPPROTO_TCP)
ipv4_found = False
for item in result:
	selftest_ipv4 = item[4][0]
	try:
		req = urllib2.Request("http://" + selftest_ipv4, headers={'User-Agent': 'SABnzbd+/1.2.3' , 'Host':'self-test.sabnzbd.org'} ) 
		f = urllib2.urlopen(req, timeout=2)    # timeout 2 seconds, in case website is not accessible
		public_ipv4 = f.read()
		socket.inet_aton(public_ipv4)  # if we got anything else than a plain IPv4 address, this will raise an exception
		# if we get here without exception, we're done:
		ipv4_found = True
		break
	except:
		# the connect OR the inet_aton raised an exception, so:
		# continue the for loop to try next server IPv4 address
		pass

if ipv4_found :
	print "IPv4 found", public_ipv4
else:
	print "No IPv4 found"
```

**Comment by Safihre** [comment]
Indeed a much more reliable solution! Make it in a PR in the function and we're all good to go :)

**Comment by thezoggy** [comment]
now your assuming the site (self-test.sabnzbd.org) itself always works..  just earlier today cloudflare was having issues.. throwing up error `HTTP/1.1 522 Origin Connection Time-out` for sites

**Comment by sanderjo** [comment]
PR sent.

**Comment by Safihre** [comment]
The other user that reported this bug on the forum now also confirms it's fixed in  RC5!

**Comment by sanderjo** [comment]
Solved in SABnzbd 1.0.0. RC5 (https://github.com/sabnzbd/sabnzbd/releases/tag/1.0.0RC5)
